apps/emqx_durable_storage/IMPLEMENTATION.md

```diff
@@ -40,36 +51,111 @@ Data format:
 `#session{clientId = "foobar", iterators = [ItKey1, ItKey2, ItKey3, ...]}`
 
 Number of records: O(N clients)
+
 Size of record: O(N subscriptions per clients)
 
 ## Iterator storage
 
 Data is written every time a client acks a message.
+
 Data is read when a client reconnects and we restart replay agents.
 
 `#iterator{key = IterKey, data = Blob}`
 
 Number of records: O(N clients * N subscriptions per client)
+
 Size of record: O(1)
+
 Write throughput: high, lots of small updates
+
 Write pattern: mostly key overwrite
+
 Read throughput: low
+
 Read pattern: random
 
 # Push vs. Pull model
 
 In push model we have replay agents iterating over the dataset in the shards.
 
-In pull model the the client processes work with iterators.
+In pull model the client processes work with iterators directly and fetch data from the remote message storage instances via remote procedure calls.
 
 ## Push pros:
 - Lower latency: message can be dispatched to the client as soon as it's persisted
 - Less worry about buffering
 
 ## Push cons:
-- Need pushback logic
+- Needs pushback logic
 - It's not entirely justified when working with external DB that may not provide streaming API
 
 ## Pull pros:
 - No need for pushback: client advances iterators at its own tempo
--
+
+## Pull cons:
+- 2 messages need to be sent over network for each batch being replayed.
```

Comment:
Q: which messages?  One requesting the iterator to advance, and its response with the payload?

Reply:
Yes, response and payload.